---
title: "Exploiting the Depmap cancer dependency data using the depmap R package"
author:
- name: Theo Killian
  affiliation: VIB-KULeuven Center for Cancer Biology, Leuven, Belgium
- name: Laurent Gatto
  affiliation: Computational Biology and Bioinformatics Unit, de Duve Institute, UCLouvain, Brussels, Belgium
date: "`r Sys.Date()`"
output: BiocWorkflowTools::f1000_article
bibliography: refs.bib
---


# Abstract

The `depmap` package facilitates access in the R environment to the
data from the Depmap project, a multi-year collaborative effort by the
Broad Institute and Wellcome Sanger Institute, mapping genetic and
chemical dependencies and other molecular biological measurements of
over 1700 cancer cell lines. The `depmap` package formats this data
for use of popular R data analysis and visualizing tools such as
`dplyr` and `ggplot2`. In addition, the `depmap` package utilizes
`ExperimentHub`, storing versions of the Depmap data accessible from
the Cloud, which may be selectively downloaded, providing a
reproducible research framework to support exploiting this data. This
paper describes a workflow demonstrating how to access and visualize
the Depmap data in R using this package.

# Keywords

cancer, cancer dependency, Depmap, ExperimentHub, data mining, 
reproducible research, Bioconductor

# Introduction

The consequences of genomic alterations of cancer cells on the molecular
biological landscape of the cell may result in differential vulnerabilities, or
"dependencies" compared to those of healthy cells. An example of genetic
dependency is a gene not necessary for the survival in healthy cells,
but essential for the vitality of particular cancer cell line. However,
the exact biological nature of many genetic dependencies of cancer are not
completely understood [@tsherniak2017defining]. A map that illustrates the
relationships between the genetic features of cancer and those of cancer
dependencies is therefore desirable. The Cancer Dependency Map or "Depmap",
a collaborative initiative between the Broad Institute and the Wellcome
Sanger Institute, aims to map genetic dependencies in a broad range of
cancer cell lines. Hundreds of cancer cell lines have been selected to be
tested in this effort, intended to mirror the distribution of various cancer
diseases in the general population. The stated aim of the Depmap Project is
developing a better understanding of the molecular biology of cancer and
the exploiting this knowledge to develop new therapies in precision cancer
medicine [@broad2020depmap].

The Depmap initative is, as of the date of this publication, an
ongoing project, with new data releases of select datasets every 90
days. As of the most current `20Q4` Depmap release, 1812 human cancer
cell lines have been mapped for dependencies [@broad2020depmap]. The
Depmap project utilizes CRISPR gene knockout as the primary method to
map genomic dependencies in cancer cell lines 
[@broad2020depmap; @meyers2017computational; @dempster2019extracting;
@dempster2019agreement]. The resulting genetic dependency score displayed
in the Depmap data is calculated from the observed log fold change in the
amount of shRNA detected in pooled cancer cell lines after gene
knockout [@cowley2014parallel; @mcfarland2018improved]. To correct
for potential off-target effects of gene knockout in overestimating
dependency with CRISPR, the Depmap iniative utilized the CERES
algorithm to moderate the final dependency score estimation
[@meyers2017computational]. It should be noted that due to advancements
in the CERES algorithm to account for CRISPR seed effects, the RNAi
dependency measurements have been rendered redundant, and further data
releases for RNAi dependency measurement have been discontinued as of the
`19Q3` release [@broad2020depmap; @dempster2019extracting]. In addition to
genomic dependency measurements of cancer cell lines, chemical dependencies
were also measured by the Depmap PRISM viability screens that as of the
`20Q4` release, tested 4,518 compounds against 578 cancer cell lines
[@corsello2019non; @broad2020depmap]. A new protemic dataset was added with
the `20Q2` release, that provides normalized quantitative profiling of
proteins of 375 cancer cell lines by mass spectrometry
[@nusinow2020quantitative]. The Depmap project has also compiled additional
datasets detailing molecular biological characterization of cancer cell
lines, such as genomic copy number, Reverse Phase Protein Array (RPPA)
data, TPM gene expression data for protein coding genes and genomic
mutation data. These datasets are updated quarterly on a release schedule
and are publically available under
[CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
licence [@broad2020depmap].

A table of the most recent datasets available for the `depmap` package
(as of `20Q4` release) is displayed in table \@ref(tab:datatab).

```{r datatab, message=FALSE, warning=FALSE, echo=FALSE, results='asis'}
library("kableExtra")
dataset_list <- read.csv("./depmap_datasets_list.csv")
datatab <- dataset_list %>%
    dplyr::select(-SourceUrl, -BiocVersion) %>%
    dplyr::filter(
               Dataset %in% c("crispr_20Q4", "rnai_19Q3", "copyNumber_20Q4",
                              "TPM_20Q4", "mutationCalls_20Q4",
                              "metadata_20Q4", " RPPA_19Q3",
                              "drug_sensitivity_19Q3", "proteomic_20Q2")) %>%
    dplyr::mutate(Dataset = sub("_.+$", "", Dataset)) %>%
    dplyr::rename(Release = SourceVersion)

knitr::kable(datatab,
             format = "latex",
             booktabs = TRUE,
             caption = paste0("Datasets available the depmap package. The ",
                              "'Release' column indicates the most recent ",
                              "available release.")) %>%
    kable_styling(full_width = TRUE,
                  latex_options = "striped",
                  font_size = 7) %>%
    column_spec(2, width = "5cm")
```


The `depmap` Bioconductor package was created in order to maximally
exploit these rich datasets and to promote reproducible research,
facilitated by importing the data into the R environment. The value
added by the  `depmap` Bioconductor package includes cleaning and
converting all datasets to long format tibbles [@muller2017tibble],
as well as adding the unique key `depmap_id` for all data tables. The
addition of the the unique key `depmap_id` aides the comparison of
molecular features and the use of common R packages such as `dplyr`
[@wickham2020package] and `ggplot2` [@wickham2011ggplot2].

As new Depmap datasets are released on a quarterly basis, it is not
feasible to include all dataset files in binary directly within the
directory of the `depmap` R package. To keep the package lightweight,
the `depmap` package utilizes and fully depends on the `ExperimentHub`
package [@ExperimentHub] to store and retrieve all versions of the
Depmap data (starting from `19Q1` through `20Q4`) in the Cloud using
AWS. The `depmap` package contains accessor functions to directly
download and cache the most current datasets from the Cloud into the
local R environment. Specific datasets (such as datasets from older releases),
which can be downloaded seperately, if desired. The `depmap` package was
designed to enhance reproducible research by ensuring datasets from all
releases will remain available in this manner. The `depmap` R package
is available as part of Bioconductor at:
[https://bioconductor.org/packages/depmap](https://bioconductor.org/packages/depmap).

# Use cases

Dependency scores are the features of primary interest in the Depmap
Project datasets. These measurements can be found in datasets `crispr`
and `rnai`, which contain information on genetic dependency, as well
as the dataset `drug_sensitivity`, which contains information pertaining
to chemical dependency. The genetic dependency can be interpreted as an
expression of how vital a particular gene for a given cancer cell line.
For example, a highly negative dependency score is derived from a large
negative log fold change in the population of cancer cells after gene
knockout or knockdown, implying that a given cell line is highly
dependent on that knocked out gene. Genes that possess highly negative
dependency scores may be interesting targets for research in cancer medicine.
In this workflow, we will describe exploring and visualizing several Depmap
datasets, including those that contain information on genetic dependency.

Below, we start by loading the packages need to run this workflow.

```{r load_libraries, message=FALSE, warning=FALSE, echo=TRUE}
library("depmap")
library("ExperimentHub")
library("dplyr")
library("ggplot2")
library("stringr")
```

The `depmap` datasets are too large to be included into a typical
package, therefore these data are stored in the Cloud. There are two
ways to access the `depmap` datasets. The first such way calls on
dedicated accessor functions that download, cache and load the latest
available dataset into the R workspace. Examples for all available
data are shown below:

```{r EH1, eval=FALSE}
rnai <- depmap_rnai()
crispr <- depmap_crispr()
copyNumber <- depmap_copyNumber()
TPM <- depmap_RPPA()
RPPA <- depmap_TPM()
metadata <- depmap_metadata()
mutationCalls <- depmap_mutationCalls()
drug_sensitivity <- depmap_drug_sensitivity()
proteomic <- depmap_proteomic()
```

Alternatively, specific dataset (from any available release) can be
accessed through Bioconductor's `ExperimentHub`. The `ExperimentHub()`
function creates an `ExperimentHub` object, which can be queried for
specific terms. The list of datasets available that correspond to the
query, `depmap` are shown below:

```{r EH2, message=FALSE, warning=FALSE, echo=TRUE}
## create ExperimentHub query object
eh <- ExperimentHub()
query(eh, "depmap")
```

Specific datasets are downloaded, cached and loaded into the workspace
as tibbles by selecting each dataset by their unique `EH` numbers.
Shown below, datasets from the `20_Q3` release are downloaded in this way.

```{r load_data, message=FALSE, warning=FALSE, echo=TRUE}
## download and cache required datasets 
crispr <- eh[["EH3797"]]
copyNumber <- eh[["EH3798"]]
TPM <- eh[["EH3799"]]
mutationCalls <- eh[["EH3800"]]
metadata <- eh[["EH3801"]]
proteomic <- eh[["EH3459"]]
```

By importing the `depmap` data into the R environment, the data can be
mined more effectively utilzing R data manipulation tools. For example,
if one was interested in researching soft tissue sarcomas and wanted to
search all such cancer cell lines for the gene with the greatest dependency,
one could accomplish this task by using functions from the `dplyr`
package. Below, the `crispr` dataset is selected for cell lines with
*"SOFT_TISSUE"* in the CCLE name, and displaying a list of the highest
dependency scores.

```{r, soft_tissue_cell_lines, echo=TRUE}
## list of dependency scores
crispr %>% 
  dplyr::select(cell_line, gene_name, dependency) %>%
  dplyr::filter(stringr::str_detect(cell_line, "SOFT_TISSUE")) %>%
  dplyr::arrange(dependency)
```

The gene `C1orf109` appears in the selected list of top dependencies scores
for soft tissue cancer cell lines. This gene, also known by the alias
Chromosome 1 Open Reading Frame 109, codes for a poorly characterized
protein which is theorized to promote cancer cell proliferation by
controlling the G1 to S phase transition [@liu2012identification]. This
protein may present as an interesting candidate target to explore and 
visualize the `depmap` data. Figure \@ref(fig:histC1orf109) displays the
`crispr` data as a histogram showing the distribution of dependency scores
for gene `C1orf109`. The red dotted line signifies the mean dependency score
for that gene, while the blue dotted line signifies the global mean dependency
score for all `crispr` measurements.

```{r histC1orf109, message=FALSE, warning=FALSE, fig.cap="Histogram of CRISPR dependency scores for gene C1orf109."}
mean_crispr_dep <- crispr %>% 
  dplyr::select(gene_name, dependency) %>% 
  dplyr::filter(gene_name == "C1orf109")

crispr %>% 
  dplyr::select(gene, gene_name, dependency) %>% 
  dplyr::filter(gene_name == "C1orf109") %>% 
  ggplot(aes(x = dependency)) + geom_histogram() +
  geom_vline(xintercept = mean(mean_crispr_dep$dependency, na.rm = TRUE), 
             linetype = "dotted", color = "red") +
  geom_vline(xintercept = mean(crispr$dependency, na.rm = TRUE),
             linetype = "dotted", color = "blue")
```

A more complex plot of the `crispr` data, is shown below \@ref(fig:C1orf109).
Visualizing this data involves plotting the distribution of dependency
scores for gene `C1orf109` for each major type of cancer, while highlighting
the qualitative nature of mutations of this gene in such cancer cell lines
(e.g. if such mutations are damaging or conserved, etc.). Notice that the
plot above \@ref(fig:histC1orf109) reflects the same overall distribution
in two dimensions.

```{r C1orf109, message=FALSE, warning=FALSE, fig.cap="Plot of CRISPR dependency scores for gene C1orf109 by lineage."}
meta_crispr <- metadata %>%
  dplyr::select(depmap_id, lineage) %>%
  dplyr::full_join(crispr, by = "depmap_id") %>%
  dplyr::filter(gene_name == "C1orf109") %>% 
  dplyr::full_join((mutationCalls %>%
                      dplyr::select(depmap_id, entrez_id, 
                                    is_cosmic_hotspot, 
                                    var_annotation)), 
                   by = c("depmap_id", "entrez_id"))

meta_crispr %>%
    ggplot(aes(x = dependency, y = lineage)) +
    geom_point(alpha = 0.4, size = 0.5) +
    geom_point(data = subset(meta_crispr,
                             var_annotation == "damaging"), 
               color = "red") +
    geom_point(data = subset(meta_crispr,
                             var_annotation == "other non-conserving"),
               color = "blue") +
    geom_vline(xintercept = mean(meta_crispr$dependency, na.rm = TRUE),
               linetype = "dotted", color = "red") +
    geom_vline(xintercept = mean(crispr$dependency, na.rm = TRUE),
               linetype = "dotted", color = "blue") 
```

Many cancer phenotypes are the result of changes in gene expression
[@li2019changes; @hernandez2019many; @felts2019stochastic]. The
extensive coverage of the `depmap` data affords visualization of
genetic expression patterns across many major types of cancer. Elevated
expression of gene `C1orf109` in lung cancer tissue has been reported in
literature [@liu2012identification]. Figure \@ref(fig:tpm) below shows a
boxplot illustrating expression values for gene `C1orf109` by lineage:

```{r tpm, message=FALSE, warning=FALSE, fig.cap="Boxplot of TPM expression values for gene C1orf109 by lineage."}
metadata %>%
  dplyr::select(depmap_id, lineage) %>%
  dplyr::full_join(TPM, by = "depmap_id") %>%
  dplyr::filter(gene_name == "C1orf109") %>%
  ggplot(aes(x = lineage, y = rna_expression, fill = lineage)) +
  geom_boxplot(outlier.alpha = 0.1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position = "none")
```

Elevated gene expression and genetic dependency in cancer cell lines
have been reported in literature [@tsherniak2017defining;
@mcfarland2018improved]. Therefore, genes with elevated gene expression
and high genetic dependency may present especially interesting
research targets. Figure \@ref(fig:rhabdomyosarcoma2)
shows a plot of expression versus CRISPR gene dependency for
Rhabdomyosarcoma. The red vertical line represents the average gene
expression for this form of cancer, while the horizontal line
represents the average dependency for this cancer type.

```{r rhabdomyosarcoma1, message=FALSE, warning=FALSE}
sarcoma <- metadata %>%
  dplyr::select(depmap_id, cell_line, primary_disease, subtype_disease) %>%
  dplyr::filter(primary_disease == "Sarcoma",
                subtype_disease == "Rhabdomyosarcoma")

crispr_sub <- crispr %>% 
  dplyr::select(depmap_id, gene, gene_name, dependency)

tpm_sub <- TPM %>% 
  dplyr::select(depmap_id, gene, gene_name, rna_expression)

sarcoma_dep <- sarcoma %>% 
  dplyr::left_join(crispr_sub, by = "depmap_id") %>%
  dplyr::select(-cell_line, -primary_disease,
                -subtype_disease, -gene_name)

sarcoma_exp <- sarcoma %>% 
  dplyr::left_join(tpm_sub, by = "depmap_id")

sarcoma_dat_exp <- dplyr::full_join(sarcoma_dep, sarcoma_exp,
                                    by = c("depmap_id", "gene")) %>%
                   dplyr::filter(!is.na(rna_expression))
```

```{r rhabdomyosarcoma2, message=FALSE, warning=FALSE, fig.cap = "Expression vs crispr gene dependency for Rhabdomyosarcoma."}
sarcoma_dat_exp %>% 
    ggplot(aes(x = dependency, y = rna_expression)) + 
    geom_point(alpha = 0.4, size = 0.5) +
    geom_vline(xintercept = mean(sarcoma_dat_exp$dependency, na.rm = TRUE),
               linetype = "dotted", color = "red") +
    geom_hline(yintercept = mean(sarcoma_dat_exp$rna_expression, na.rm = TRUE),
               linetype = "dotted", color = "red") +
    theme(axis.text.x = element_text(angle = 45))
```

Genes with the highest depenency scores and highest TPM gene expression
are found in the upper left section of the plot above
\@ref(fig:rhabdomyosarcoma2). Almost all of the genes with the
highest depenency scores display increased expression.

```{r}
sarcoma_dat_exp %>%
  dplyr::select(cell_line, gene_name, dependency, rna_expression) %>%
  dplyr::arrange(dependency, rna_expression)
```

Changes in genomic copy number may also play a role in some cancer
phenotypes [@meyers2017computational; @aguirre2016genomic;
@shao2019copy]. The `depmap` data allows the display of log genomic
copy number for across many cancer lineages. Figure
\@ref(fig:boxC1orf109) shows such a plot for gene `C1orf109` for each major type
of cancer lineage:

```{r boxC1orf109, message=FALSE, warning=FALSE, fig.cap="Boxplot of log copy number for gene C1orf109 by lineage"}
metadata %>%
  dplyr::select(depmap_id, lineage) %>%
  dplyr::full_join(copyNumber, by = "depmap_id") %>%
  dplyr::filter(gene_name == "C1orf109") %>%
  ggplot(aes(x = lineage, y = log_copy_number, fill = lineage)) +
  geom_boxplot(outlier.alpha = 0.1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position = "none")
```

# Discussion and outlook

We hope that this package will be used by cancer researchers to dig deeper into
the Depmap data and to support their research. Additionally, we highly encourage
future `depmap` users to combine `depmap` data with other datasets of interest,
such as `TCGA` and `CCLE`.

The depmap R package will continue to be maintained in line with the biannual
Bioconductor release, in addition to quarterly releases of Depmap data.

We welcome feedback and questions from the community. We also highly appreciate
contributions to the code in the form of pull requests.

# Software availability

All packages used in this workflow are available from the
Comprehensive R Archive Network (https://cran.r-project.org) or
Bioconductor (http://bioconductor.org). The specific version numbers
of R and the packages used are shown below.

Note, the latest `depmap` datasets are only available through
`ExperimentHub` by installing and running the latest `devel` version
of Bioconductor. To install the `devel` version of Bioconductor:

```{r message=FALSE, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "devel")
BiocManager::valid()              # checks for out of date packages
```

To install the `depmap` package:

```{r message=FALSE, eval=FALSE}
BiocManager::install('depmap')
```

```{r echo = FALSE}
sessionInfo()
```

# Acknowledgements

# Competing interests

No competing interests were disclosed.

# Grant information

# References

